components/esp_http_server/include/esp_http_server.h

```diff
@@ -1452,6 +1458,66 @@ esp_err_t httpd_queue_work(httpd_handle_t handle, httpd_work_fn_t work, void *ar
  * @}
  */
 
+/* ************** Group: WebSocket ************** */
+/** @name WebSocket
+ * Functions and structs for WebSocket server
+ * @{
+ */
+
+/**
+ * @brief Enum for WebSocket packet types (Opcode in the header)
+ * @note Please refer to RFC6455 Section 5.4 for more details
+ */
+typedef enum {
+    HTTPD_WS_TYPE_CONTINUE   = 0x0,
+    HTTPD_WS_TYPE_TEXT       = 0x1,
+    HTTPD_WS_TYPE_BINARY     = 0x2,
+    HTTPD_WS_TYPE_CLOSE      = 0x8,
+    HTTPD_WS_TYPE_PING       = 0x9,
+    HTTPD_WS_TYPE_PONG       = 0xA
+} httpd_ws_type_t;
+
+/**
+ * @brief WebSocket packet format
+ */
+typedef struct httpd_ws_pkt {
+    bool final;                 /*!< Final packet */
+    httpd_ws_type_t type;       /*!< WebSocket packet type */
+    uint8_t *payload;           /*!< Pre-allocated data buffer */
+    size_t len;                 /*!< Length of the WebSocket data */
+} httpd_ws_pkt_t;
+
+/**
+ * @brief Receive and parse a WebSocket packet
+ * @param[in]   req         Current request
+ * @param[out]  pkt         WebSocket packet
+ * @param[in]   max_len     Maximum length for receive
+ * @param[in]   timeout_sec Timeout in second(s), leave it to 0 for forever
+ * @return
+ *  - ESP_OK                    : On successful
+ *  - ESP_FAIL                  : Socket errors occurs
+ *  - ESP_ERR_INVALID_STATE     : Handshake was already done beforehand
+ *  - ESP_ERR_INVALID_ARG       : Argument is invalid (null or non-WebSocket)
+ */
+esp_err_t httpd_ws_recv_pkt(httpd_req_t *req, httpd_ws_pkt_t *pkt, size_t max_len, uint32_t timeout_sec);
+
+/**
+ * @brief Construct and send a WebSocket packet
+ * @param[in]   req     Current request
+ * @param[in]   pkt     WebSocket packet
+ * @return
+ *  - ESP_OK                    : On successful
+ *  - ESP_FAIL                  : When socket errors occurs
+ *  - ESP_ERR_INVALID_STATE     : Handshake was already done beforehand
+ *  - ESP_ERR_INVALID_ARG       : Argument is invalid (null or non-WebSocket)
+ */
+esp_err_t httpd_ws_send_pkt(httpd_req_t *req, httpd_ws_pkt_t *pkt);
```

Comment:
@shahpiyushv I think I can, but it might be a bit too complex. I think this PR is just for basic WebSocket server support. It should be considered in the later PRs.

However, if you really think I should include this async support anyway, I will do it in this PR.